lib/Service/ImageService.php

```diff
@@ -0,0 +1,676 @@
+		$saveDir = $this->getOrCreateAttachmentDirectoryForFile($textFile);
+		if ($saveDir !== null) {
+			return $this->downloadLink($saveDir, $link, $textFile);
+		} else {
+			return [
+				'error' => 'Impossible to get attachment directory',
+			];
+		}
+	}
+
+	/**
+	 * Download and save an image from a link in the attachment folder in a public context
+	 *
+	 * @param int|null $documentId
+	 * @param string $link
+	 * @param string $shareToken
+	 * @return array|string[]
+	 * @throws Exception
+	 */
+	public function insertImageLinkPublic(?int $documentId, string $link, string $shareToken): array {
+		if (!$this->hasUpdatePermissions($shareToken)) {
+			throw new Exception('No write permissions');
+		}
+		$textFile = $this->getTextFilePublic($documentId, $shareToken);
+		$saveDir = $this->getOrCreateAttachmentDirectoryForFile($textFile);
+		if ($saveDir !== null) {
+			return $this->downloadLink($saveDir, $link, $textFile);
+		} else {
+			return [
+				'error' => 'Impossible to get attachment directory',
+			];
+		}
+	}
+
+	/**
+	 * Check if the shared access has write permissions
+	 *
+	 * @param string $shareToken
+	 * @return bool
+	 */
+	private function hasUpdatePermissions(string $shareToken): bool {
+		try {
+			$share = $this->shareManager->getShareByToken($shareToken);
+			return ($share->getShareType() === IShare::TYPE_LINK && $share->getPermissions() & Constants::PERMISSION_UPDATE);
+		} catch (ShareNotFound $e) {
+			return false;
+		}
+	}
+
+	/**
+	 * Download an image from a link and place it in a given folder
+	 * @param Folder $saveDir
+	 * @param string $link
+	 * @param File $textFile
+	 * @return array
+	 * @throws NotFoundException
+	 * @throws \OCP\Files\InvalidPathException
+	 * @throws \OCP\Files\NotPermittedException
+	 * @throws \OCP\Lock\LockedException
+	 */
+	private function downloadLink(Folder $saveDir, string $link, File $textFile): array {
+		$fileName = (string) time();
+		$savedFile = $saveDir->newFile($fileName);
+		$resource = $savedFile->fopen('wb');
+		$res = $this->simpleDownload($link, $resource);
+		if (is_resource($resource)) {
+			fclose($resource);
+		}
+		$savedFile->touch();
+		if (isset($res['Content-Type'])) {
+			if (in_array($res['Content-Type'], ImageController::IMAGE_MIME_TYPES)) {
+				if ($res['Content-Type'] === 'image/jpeg') {
+					$fileName = $fileName . '.jpg';
+				} elseif ($res['Content-Type'] === 'image/x-xbitmap' || $res['Content-Type'] === 'image/x-ms-bmp') {
+					$fileName = $fileName . '.bmp';
+				} elseif ($res['Content-Type'] === 'image/svg+xml') {
+					$fileName = $fileName . '.svg';
+				} else {
+					$ext = preg_replace('/^image\//i', '', $res['Content-Type']);
+					$fileName = $fileName . '.' . $ext;
+				}
+				$targetPath = $saveDir->getPath() . '/' . $fileName;
+				$savedFile->move($targetPath);
+				// get file type and name
+				return [
+					'name' => $fileName,
+					'id' => $savedFile->getId(),
+					'documentId' => $textFile->getId(),
+				];
+			} else {
```

Comment:
From PhpStorm: [EA] Child instructions can be extracted here (clearer intention, lower complexity numbers).